[SPARC][MC] Fix `%gdop_hix22()` and `%gdop_lox10()` to use correct relocations

alexrp opened the pull request:
1bfc5e7 introduced support for `%gdop_hix22()` and `%gdop_lox10()`. However, it incorrectly mapped them to `R_SPARC_GOTDATA_HIX22` and `R_SPARC_GOTDATA_LOX10`. They should in fact emit `R_SPARC_GOTDATA_OP_HIX22` and `R_SPARC_GOTDATA_OP_LOX10`.

This became a problem when assembling glibc's PIC startup code:

```asm
sethi %gdop_hix22(main), %o0
xor %o0, %gdop_lox10(main), %o0
ldx [%l7 + %o0], %o0, %gdop(main)
```

After the `xor`, `%o0` should contain the GOT offset for `main`, but because of the incorrect relocations, it actually ends up containing the address of `main`, which of course makes the following `ldx` fail.

alexrp commented:
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️

This would just be pointless formatting churn in the existing code.

MaskRay commented:
> > ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> This would just be pointless formatting churn in the existing code.

You might add `// clang-format off ` and on if needed, like what I added to SparcMCExpr.cpp

Note: if a fixup kind always leads to a relocation, you can encode its value literally. See
```
% rg R_LARCH_TLS_IE_PC_HI20 llvm/lib/
llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchMCCodeEmitter.cpp
214:      FixupKind = ELF::R_LARCH_TLS_IE_PC_HI20;
```
and llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp for examples.

alexrp commented:
> You might add `// clang-format off ` and on if needed, like what I added to SparcMCExpr.cpp

Done.

> Note: if a fixup kind always leads to a relocation, you can encode its value literally.

That makes sense, but from what I can tell, the SPARC backend would need some refactoring to do this, so probably best not to mix in with this change?

koachan reviewed (approved):
Other than those already raised by maskray, I think this seems okay~

s-barannikov reviewed (approved):
LGTM

alexrp commented on llvm/lib/Target/Sparc/MCTargetDesc/SparcFixupKinds.h:
```diff
@@ -100,16 +101,17 @@ namespace llvm {
       LastTargetFixupKind,
       NumTargetFixupKinds = LastTargetFixupKind - FirstTargetFixupKind
     };
+  // clang-format off
```
Oops, good catch.

MaskRay commented on llvm/lib/Target/Sparc/MCTargetDesc/SparcAsmBackend.cpp:
```diff
@@ -202,9 +203,9 @@ namespace {
-        { "fixup_sparc_gotdata_hix22",  0,  0,  0 },
-        { "fixup_sparc_gotdata_lox10",  0,  0,  0 },
-        { "fixup_sparc_gotdata_op",     0,  0,  0 },
+        { "fixup_sparc_gotdata_op_hix22", 10, 22,  0 },
```
For fixup kinds that always map to relocation types (e.g. GOT-generating and TLS relocation types), the offset/length fields can be kept as 0. The only difference is in the -filetype=asm -show-encoding output, which is not really relevant. (We should test relocations using -filetype=obj; the -filetype=asm -show-encoding output is informal.)

MaskRay replied:
If you rebase after 7d423c09c0e063eeeaa2f91100eb6f04dc15767c , you can even remove the fixup kinds.

alexrp replied:
Sounds good, I'll try rebasing on that.

alexrp committed:
[SPARC][MC] Fix %gdop_hix22() and %gdop_lox10() to use correct relocations.

1bfc5e7 introduced support for %gdop_hix22() and %gdop_lox10(). However, it
incorrectly mapped them to R_SPARC_GOTDATA_HIX22 and R_SPARC_GOTDATA_LOX10.
They should in fact emit R_SPARC_GOTDATA_OP_HIX22 and R_SPARC_GOTDATA_OP_LOX10.

This became a problem when assembling glibc's PIC startup code:

    sethi %gdop_hix22(main), %o0
    xor %o0, %gdop_lox10(main), %o0
    ldx [%l7 + %o0], %o0, %gdop(main)

After the xor, %o0 should contain the GOT offset for main, but because of the
incorrect relocations, it actually ends up containing the address of main, which
of course makes the following ldx fail.
